### PR TITLE
Add Jest tests for api

### PR DESCRIPTION
**Pull request by robmosca**
While working on the import-map-overrides project, I realized the code is not covered by unit tests. I would like to contribute Jest tests, to make sure that further modifications to the package do not introduce regressions.

Current coverage:

Screenshot 2023-07-16 at 12 44 37

### CODE REVIEW

File: jest.setup.js [hunk 1]
```diff
@@ -0,0 +1,4 @@
+import { TextEncoder, TextDecoder } from "util";
+
+global.TextEncoder = TextEncoder;
```
Comment (joeldenning) [inline jest.setup.js:3]: What is this needed for?

File: package.json [hunk 2]
```diff
@@ -12,6 +12,7 @@
     "clean": "rimraf dist",
     "watch": "cross-env NODE_ENV=development rollup -c --watch",
     "copy-test-files": "copyfiles test/**/* dist -f",
+    "test-jest": "jest",
     "test-serve": "serve .",
     "test-embed-serve": "serve . -l 3333",
     "test": "concurrently pnpm:watch pnpm:copy-test-files pnpm:test-serve pnpm:test-embed-serve",
```
Comment (joeldenning) [inline package.json:18]: The `test` command should run the automated tests now that we have them.

```suggestion
    "test": "jest",
```

File: package.json [hunk 3]
```diff
@@ -12,6 +12,7 @@
     "clean": "rimraf dist",
     "watch": "cross-env NODE_ENV=development rollup -c --watch",
     "copy-test-files": "copyfiles test/**/* dist -f",
+    "test-jest": "jest",
```
Comment (joeldenning) [inline package.json:15]: Let's make the jest command be the `"test"` script

```suggestion
```

File: package.json [hunk 4]
```diff
@@ -12,6 +12,7 @@
...
     "test-serve": "serve .",
```
Comment (joeldenning) [inline package.json:16]: Rename to `serve`

```suggestion
    "serve": "serve .",
```

File: package.json [hunk 5]
```diff
@@ -12,6 +12,7 @@
...
     "test-embed-serve": "serve . -l 3333",
```
Comment (joeldenning) [inline package.json:17]:
```suggestion
    "embed-serve": "serve . -l 3333",
```

File: src/api/js-api.js [hunk 6]
```diff
@@ -210,11 +210,9 @@ function init() {
                 nextPromise = Promise.resolve(JSON.parse(scriptEl.textContent));
               }
 
-              return Promise.all([
-                promise,
-                nextPromise,
-              ]).then(([originalMap, newMap]) =>
-                imo.mergeImportMap(originalMap, newMap)
+              return Promise.all([promise, nextPromise]).then(
```
Comment (joeldenning) [inline src/api/js-api.js:213]: Are these prettier changes because of your installation of VS Code? Since the version of prettier is pinned, I don't know why these changes are happening. Please add a `check-format` script to the package.json that runs `prettier --check .` What I want to avoid is a situation where a global installation of prettier that's specific to one contributor's computer is changing the format of the code

File: .github/workflows/build-and-test.yml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+      - uses: actions/setup-node@v2
+        with:
+          node-version: "16"
+      - uses: pnpm/action-setup@v2.0.1
+        with:
+          version: 8.4.0
+      - run: git checkout main && git checkout $GITHUB_SHA
+      - run: pnpm install --frozen-lockfile
+      - run: pnpm run build
+      - run: pnpm run lint
```
Comment (joeldenning) [inline .github/workflows/build-and-test.yml:26]: Please add a check-format script to the package.json to verify prettier format (`prettier --check .`). I generally don't use eslint for formatting, so it's a separate command:

```suggestion
      - run: pnpm run lint
      - run: pnpm run check-format
```

File: jest.setup.js (hunk 1 above)
Comment (robmosca) [reply jest.setup.js:3]: Actually, it was needed in one of the approaches I tried for the testing, but it's not needed anymore. I'll remove it.

File: src/api/js-api.js (hunk 6 above)
Comment (robmosca) [reply src/api/js-api.js:213]: See https://github.com/single-spa/import-map-overrides/pull/81/files#r1248903707

**Comment by robmosca** [comment]
Addressed the initial comments. I will continue working on adding tests.

File: jest.setup.js (hunk 1 above)
Comment (robmosca) [reply jest.setup.js]: Removed.

File: package.json (hunk 5 above)
Comment (robmosca) [reply package.json]: Modified

File: jest.setup.js [hunk 8]
```diff
@@ -0,0 +1 @@
+require("jest-fetch-mock").enableMocks();
```
Comment (robmosca) [inline jest.setup.js:1]: This is needed in order to safely mock fetch (for testing external maps)

File: src/api/js-api.js [hunk 9]
```diff
@@ -207,14 +207,18 @@ function init() {
...
                 nextPromise = fetchExternalMap(scriptEl.src);
               } else {
-                nextPromise = Promise.resolve(JSON.parse(scriptEl.textContent));
+                let parsedContent;
+                try {
+                  parsedContent = JSON.parse(scriptEl.textContent);
+                } catch (e) {
+                  parsedContent = createEmptyImportMap();
+                }
+                nextPromise = Promise.resolve(parsedContent);
```
Comment (robmosca) [inline src/api/js-api.js:216]: While testing I realized this function bombs when the import map is empty or is malformed. To prevent a malformed import map from breaking the plugin, I modified this code to instead return an empty map.

File: src/api/js-api.js [hunk 10]
```diff
@@ -342,11 +345,10 @@ function init() {
...
         externalOverrideMapPromises[importMapUrl] ||
-        (externalOverrideMapPromises[importMapUrl] = fetchExternalMap(
-          importMapUrl
-        ));
-      return promise.then(() =>
-        includes(imo.invalidExternalMaps, importMapUrl)
+        (externalOverrideMapPromises[importMapUrl] =
+          fetchExternalMap(importMapUrl));
+      return promise.then(
+        () => !includes(imo.invalidExternalMaps, importMapUrl)
```
Comment (robmosca) [inline src/api/js-api.js:351]: I believe this was a bug. I found it out when writing a test for this function. The map is valid if it IS NOT included in the `invalidExternalMaps` array.

**Comment by robmosca** [comment]
These changes are included in https://github.com/single-spa/import-map-overrides/pull/89 and https://github.com/single-spa/import-map-overrides/pull/94.
If you prefer reviewing the most comprehensive PR (https://github.com/single-spa/import-map-overrides/pull/94) we can close this one.

**Comment by robmosca** [comment]
Closed in favour of https://github.com/single-spa/import-map-overrides/pull/96
